### ISSUE TITLE
Bug: Can't use bigint for the key property

### ISSUE THREAD

**Issue by zmrocze** [facebook/react#27351]
React version: 18.2.0

## Steps To Reproduce

1. Use bigint as the key property of a jsx object: `<Div key={4n}`

## The current behavior

Observe a type error, because the accepted type is: `number | string | undefined | null`.

## The expected behavior

The type `bigint` is acceptable type for the key property.

Come on guys: really floats are keys but ints are not?

**Comment by MarceloSpessoto** [comment]
Hello, could you give a code example of the bug happening?

**Comment by zmrocze** [comment]
the bug is simpler than the template expected, sorry if I confused.

The problem is the definition of the `Key` type. The type doesn't include `bigint`s BUT imo it should.

So really, what reproduction should I show more than the single line that's already there?

**Comment by rizfanradya** [comment]
when you try to create a BigInt like BigInt(4), it won't result in a proper BigInt. Instead, it will produce a regular Number data type. You should use a different notation to create BigInt, such as adding "n" at the end of the number:

```
const bigIntValue = BigInt(4); // This will produce a Number value, not BigInt
const bigIntValueCorrect = 4n; // This will produce a BigInt

```
So, to create a BigInt with the value 4, you should use the notation 4n, not BigInt(4). BigInt values created using the 4n notation will have very high precision and are suitable for performing operations with large numbers that cannot be represented by regular Number data type.

**Comment by zmrocze** [comment]
@rizfanradya Ah, thank you, that's good info as I'm novice with ts.

The problem description stands as is, I've just fixed the code example.

**Comment by sidharth8338** [comment]
Use like this `<Div key={4n.toString()} />`
By converting the `BigInt` to a `string`, you can use it as a key without encountering type errors. However, keep in mind that React will treat these keys as strings, which should work for most scenarios.

**Comment by LucaDillenburg** [comment]
I can reproduce the issue. The following code gives this error: "Type 'bigint' is not assignable to type 'Key | null | undefined'"
`<div className="App" key={4n}>`

I will work with @MarceloSpessoto on this.

**Comment by LucaDillenburg** [comment]
@zmrocze, after analysing the code with @MarceloSpessoto, we've identified that the Key type definition is not present in this repository. Essential React types are typically defined within the [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) repository. More specifically, the Key type is defined in [this file](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/babel-plugin-react-html-attrs/index.d.ts).

Nonetheless, we've taken the initiative to create two Pull Requests aimed at introducing support for bigints in keys. You can access these Pull Requests here:
React - [PR]()
DefinitelyTyped - [PR]()

In the meantime, we recommend using a string with `key={4n.toString()}` as a workaround.

**Comment by zmrocze** [comment]
@LucaDillenburg Hey, thats great! Thanks

**Comment by sebmarkbage** [comment]
React's runtime have slightly broader implementation than it's TypeScript types because it is possible to use it without types and it's also possible to make mistakes with the types. E.g. you can pass all kinds of values to DOM nodes and they will typically be converted to a string. This lets you pass things like objects etc. That's just common in JavaScript.

React runtime already supports bigint being passed to the key because it'll convert it to a string key. So it behaves as its string value. That's also how numbers work.

The TypeScript definitions are sometimes more restrictive than the runtime to catch common mistakes. So I think the ask here is just to update the TypeScript definition to allow it. Note that it'll still be treated as a string though.

**Comment by eps1lon** [comment]
We have a lot of places where React (or JS) implicitly converts values to strings. TypeScript generally does not allow this at the type level and wants an explicit conversion. So we've followed that model for consistency. 

Allowing `bigint` is not much different than allowing `{ toString(): string }`. But that we wouldn't want because we can't make any inference about how semantically useful the `toString` method would be for keys.

Allowing `bigint` specifically makes sense since we already do the same for `number`. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66723 was already a good start but it should also make the same change to `types/react/ts5.0/index.d.ts`. Feel free to open a PR making these two changes.

**Comment by MarceloSpessoto** [comment]
Ok @eps1lon , I will make these changes.

**Comment by MarceloSpessoto** [comment]
Done

**Comment by bombillazo** [comment]
This has caused some issues for use when using Keys in places bigint is not supported:
https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/66749

**Comment by eps1lon** [comment]
Fixed in `@types/react@18.2.22`. 

The breakage happens in places where `React.Key` was used for something it used to structurally match but not semantically so this seems fine with me. `React.Key` is the type `key` accepts, not a type you can key arbitrary objects with.
